x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/logical/Not.java

```diff
@@ -60,15 +60,23 @@ public String processScript(String script) {
 
     @Override
     protected Expression canonicalize() {
-        Expression canonicalChild = field().canonical();
-        if (canonicalChild instanceof Negatable) {
-            return ((Negatable) canonicalChild).negate();
+        if (field() instanceof Negatable) {
+            return ((Negatable) field()).negate().canonical();
```

Comment:
Canonicalize after negation to avoid recreating the trees twice since canonical already handles normalization and ordering.